File: 02-LINKED-LISTS/design-browser-history.py
from typing import *
from ListNode import ListNode

# https://leetcode.com/problems/design-browser-history/
class BrowserHistory:
    def __init__(self, homepage: str):
        self.current = ListNode(homepage)

    def visit(self, url: str) -> None:
        visitedNode = ListNode(url)
        visitedNode.prev = self.current
        self.current.next,  = visitedNode
        self.current = visitedNode

    def back(self, steps: int) -> str:
        while steps > 0 and self.current.prev:
            self.current = self.current.prev
            steps -= 1
        return self.current.val

    def forward(self, steps: int) -> str:
        while steps > 0 and self.current.next:
            self.current = self.current.next
            steps -= 1
        return self.current.val

File: 02-LINKED-LISTS/PHP/design-browser-history.php
<?php

# https://leetcode.com/problems/design-browser-history/
class BrowserHistory
{
    private ListNode $current;

    public function __construct(string $homepage)
    {
        $this->current = new ListNode($homepage);
    }

    public function visit(string $url): void
    {
        $visitedNode = new ListNode($url);
        $visitedNode->prev = $this->current;
        $this->current->next = $visitedNode;
        $this->current = $visitedNode;
    }

    public function back(int $steps): string
    {
        while ($steps > 0 && $this->current->prev) {
            $this->current = $this->current->prev;
            $steps--;
        }
        return $this->current->val;
    }

    public function forward(int $steps): string
    {
        while ($steps > 0 && $this->current->next) {
            $this->current = $this->current->next;
            $steps--;
        }
        return $this->current->val;
    }
}